frontend/userpatch.lua

```diff
@@ -0,0 +1,138 @@
+        table.insert(command, "'")
+        table.insert(command, tostring(v))
+        table.insert(command, "' ")
+    end
+    local cmd = table.concat(command)
+    logger.info("Live update apply:", cmd)
+    local retval =  os.execute(cmd)
+    logger.info("Live update script returned:", retval)
+end
+
+--- Run user shell scripts or lua patches
+-- Execution order order is alphanum-sort for humans version 4: `1-patch.lua` is executed before `10-patch.lua`
+-- (see http://notebook.kulchenko.com/algorithms/alphanumeric-natural-sorting-for-humans-in-lua)
+-- string directory ... to scan through (flat no recursion)
+-- string priority ... only files starting with `priority` followed by digits and a '-' will be processed.
+-- bool update_once_pending ... if true the onetime update has not been executed
+-- string update_once_marker ... remove this file, if onetime update gets applied the first time
+local function runLiveUpdateTasks(dir, priority, update_once_pending, update_once_marker)
+    if lfs.attributes(dir, "mode") ~= "directory" then
+        return
+    end
+
+    local patches = {}
+    for entry in lfs.dir(dir) do
+        local mode = lfs.attributes(dir .. "/" .. entry, "mode")
+        if entry and mode == "file" and entry:match("^" .. priority .. "%d*%-") then
+            table.insert(patches, entry)
+        end
+    end
+
+    if #patches == 0 then
+        return -- nothing to do
+    end
+
+    local function addLeadingZeroes(d)
+        local dec, n = string.match(d, "(%.?)0*(.+)")
+        return #dec > 0 and ("%.12f"):format(d) or ("%s%03d%s"):format(dec, #n, n)
+    end
+    local sorting = function(a, b)
+        return tostring(a):gsub("%.?%d+", addLeadingZeroes)..("%3d"):format(#b)
+            < tostring(b):gsub("%.?%d+", addLeadingZeroes)..("%3d"):format(#a)
+    end
+
+    table.sort(patches, sorting)
+
+    for i, entry in ipairs(patches) do
+        local fullpath = dir .. "/" .. entry
+        if lfs.attributes(fullpath, "mode") == "file" then
+            if fullpath:match("%.lua$") then -- execute patch-files first
+                logger.info("Live update apply:", fullpath)
+                local ok, err = pcall(dofile, fullpath)
+                if not ok then
+                    logger.warn("Live update", err)
+                    if priority >= userpatch.late then -- Only show InfoMessage, when late during startup.
+                        -- Only developers (advanced users) will use this mechanism.
+                        -- A warning on a patch failure after an OTA update will simplify troubleshooting.
+                        local UIManager = require("ui/uimanager")
+                        local InfoMessage = require("ui/widget/infomessage")
+                        UIManager:show(InfoMessage:new{text = "Error loading patch:\n" .. fullpath}) -- no translate
+                    end
+                end
+            elseif fullpath:match("%.sh$") then -- execute shell scripts secont
```

Comment:
```
-- execute patch-files first
-- execute shell scripts secont (second)
```
Any reason for this ordering ?
I guess with the earlier ordering, one could still have sh scripts run before lua ones, with
```
12-my_first_script.sh
13-my_then_script.lua
```
right ?

Reply:
The order is for example
```
1-patch.lua
10-patch.lua
11-patch.lua
11-patch.sh
12-patch.sh
13-patch.lua
14-patch.sh
100-patch.sh
101-patch.lua
101-patch.sh
....
100- .....
```

If two names are the same and only the extension is different `lua` gets executed before `sh` (alphabetically).